{"metadata": {"language_info": {"file_extension": ".py", "codemirror_mode": {"version": 3, "name": "ipython"}, "name": "python", "nbconvert_exporter": "python", "mimetype": "text/x-python", "version": "3.7.3", "pygments_lexer": "ipython3"}, "widgets": {"state": {"dc62ee2ed17b4b7ab96a5e4b73d6b521": {"views": [{"cell_index": 10}]}, "52c44c2ff34e4cf69825b4e9ba8ac674": {"views": [{"cell_index": 7}]}, "9b98374208954f2bb68b6e45664d0e9e": {"views": [{"cell_index": 4}]}, "2a5e839a40ac4aad9e170213613fb672": {"views": [{"cell_index": 1}]}}, "version": "1.2.0"}, "anaconda-cloud": {}, "genepattern": {"repository_url": "http://notebook.genepattern.org/services/sharing/notebooks/337/"}, "kernelspec": {"language": "python", "display_name": "Python 3.7", "name": "python3.7"}}, "cells": [{"metadata": {}, "source": ["# Differential Expression: RNA-Seq data in GenePattern Notebook"], "cell_type": "markdown"}, {"metadata": {}, "source": ["Compute differentially expressed genes or transcripts and visualize the results"], "cell_type": "markdown"}, {"metadata": {}, "source": ["<div class=\"alert alert-info\">\n", "<h3 style=\"margin-top: 0;\"> Instructions <i class=\"fa fa-info-circle\"></i></h3>\n", "<ul><li>Sign in to GenePattern by entering your username and password into the form below. </li></ul>\n", "</div>"], "cell_type": "markdown"}, {"metadata": {"genepattern": {"type": "auth", "server": "https://cloud.genepattern.org/gp", "name": "Login"}}, "source": ["# Requires GenePattern Notebook: pip install genepattern-notebook\n", "import gp\n", "import genepattern\n", "\n", "# Username and password removed for security reasons.\n", "genepattern.display(genepattern.session.register(\"https://cloud.genepattern.org/gp\", \"\", \"\"))"], "cell_type": "code", "outputs": [{"data": {"text/plain": ["GPAuthWidget()"], "application/vnd.jupyter.widget-view+json": {"version_major": 2, "model_id": "640a5e8ded6a4bc58042e20a4461cc8c", "version_minor": 0}}, "metadata": {}, "output_type": "display_data"}], "execution_count": 7}, {"metadata": {}, "source": ["# 1. Differential expression: tumor vs normal"], "cell_type": "markdown"}, {"metadata": {}, "source": ["<p>Below you can find the files we will use for differential expression.</p>\n", "\n", "<ul>\n", "    <li><a class=\"nbtools-markdown-file\" href=\"https://datasets.genepattern.org/data/ccmi_tutorial/2018-09-20/BRCA_HUGO_symbols.preprocessed.gct\">BRCA_HUGO_symbols.preprocessed.gct</a></li>\n", "    <li><a class=\"nbtools-markdown-file\" href=\"https://datasets.genepattern.org/data/ccmi_tutorial/2018-09-20/BRCA_Dataset.cls\">BRCA_Dataset.cls</a></li>\n", "</ul>"], "cell_type": "markdown"}, {"metadata": {}, "source": ["## Compute differentially expressed genes\n"], "cell_type": "markdown"}, {"metadata": {}, "source": ["<p>The following analysis cell uses an information-theoretic method to find significantly differentially expressed transcripts or genes.</p>\n", "\n", "<!--<p>--&gt; If you are using microarray data you can use ComparativeMarkerSelection, or DESeq2 as an alternative...</p>-->\n", "\n", "<div class=\"alert alert-info\">\n", "<h3 style=\"margin-top: 0;\"> Instructions <i class=\"fa fa-info-circle\"></i></h3>\n", "<ol>\n", "    <li>For the <strong>gene expression dataset</strong> parameter, select from the dropdown menu the file:&nbsp;<strong>BRCA_dataset.preprocessed.gct</strong></li>\n", "\t<li>For the <strong>phenotype file</strong> parameter,&nbsp;select from the dropdown menu the file:&nbsp;<strong>BRCA_Dataset.cls</strong></li>\n", "\t<li>Leave the rest of the parameters as default.\n", "    <!-->\n", "\t<ul>\n", "\t\t<li><strong>output filename</strong>: diffex_output</li>\n", "\t\t<li><strong>ranking method</strong>: Pearson Correlation Matching</li>\n", "\t\t<li><strong>max number of genes to show</strong>: 20</li>\n", "\t\t<li><strong>number of permutations</strong>: 10</li>\n", "\t\t<li><strong>title</strong>: Differential Expression Results</li>\n", "\t\t<li><strong>random seed</strong>: 20180920</li>\n", "\t\t<li><strong>output variable</strong>: gene_scores</li>\n", "\t</ul>\n", "    </!-->\n", "\t</li>\n", "\t<li>Click <strong>Run</strong></li>\n", "</ol>\n", "</div>\n"], "cell_type": "markdown"}, {"metadata": {"nbtools": {"show_code": false, "description": "Sort genes according to their association with a discrete phenotype or class vector.", "type": "uibuilder", "param_values": {"number_of_permutations": "10", "output_var": "gene_scores", "phenotype_file": [], "output_filename": "diffex_output", "random_seed": "20180920", "ranking_method": "custom_pearson_corr", "gene_expression": [], "max_number_of_genes_to_show": "20", "title": "Differential Expression Results", "show_plot": "true"}, "name": "Differential gene expression, discrete pheotype."}, "genepattern": {"type": "uibuilder", "output_variable": "gene_scores", "param_values": {"number_of_permutations": "30", "output_var": "gene_scores", "phenotype_file": [], "output_filename": "diffex_output", "random_seed": "20180314", "ranking_method": "custom_pearson_corr", "gene_expression": [], "max_number_of_genes_to_show": "20", "title": "Differential Expression Results"}, "show_code": false}, "scrolled": true}, "source": ["import genepattern\n", "import ccalnoir as ccal\n", "\n", "from ccalnoir import differential_gene_expression\n", "import pandas as pd\n", "import numpy as np\n", "from ccalnoir import compute_information_coefficient\n", "from ccalnoir import custom_pearson_corr\n", "RANDOM_SEED = 20180314\n", "\n", "genepattern.GPUIBuilder(differential_gene_expression, name=\"Differential gene expression, discrete pheotype.\", \n", "                        description=\"Sort genes according to their association with a discrete phenotype or class vector.\",\n", "                        parameters={\n", "                                \"gene_expression\":{\"name\":\"gene_expression_dataset\",\n", "                                                   \"type\": \"file\",\n", "                                                   \"kinds\": [\"gct\"]},\n", "                                \"phenotype_file\":{\"type\": \"file\",\n", "                                                  \"kinds\": [\"cls\"]},\n", "                                \"ranking_method\":{\n", "                                                  \"default\": \"custom_pearson_corr\",\n", "                                                  \"choices\":{'Pearson Correlation Matching':\"custom_pearson_corr\",\n", "                                                             'Information Coefficient Matching':\"compute_information_coefficient\",\n", "                                                            }\n", "                                                     },\n", "                                \"title\":{\"default\":\"Differential Expression Results\"},\n", "                                \"output_filename\":{\"default\": \"diffex_output\",},\n", "                                \"ramdon_seed\":{\"default\":20180314,},\n", "                                \"output_var\":{\"default\": \"gene_scores\",},\n", "                        })"], "cell_type": "code", "outputs": [{"data": {"text/plain": ["UIBuilder(description='Sort genes according to their association with a discrete phenotype or class vector.', \u2026"], "application/vnd.jupyter.widget-view+json": {"version_major": 2, "model_id": "fbb3d00935f24fb68625f89fd6bb84aa", "version_minor": 0}}, "metadata": {}, "output_type": "display_data"}], "execution_count": 2}, {"metadata": {}, "source": ["## Investigate differentially expressed genes"], "cell_type": "markdown"}, {"metadata": {}, "source": ["<p>For a quick look at each of the top differentially expressed genes we probe <a href=\"http://www.genecards.org/\" target=\"_blank\">http://www.genecards.org/</a> using the Table Creator function</p>\n", "\n", "<div class=\"alert alert-info\">\n", "<h3 style=\"margin-top: 0;\"> Instructions <i class=\"fa fa-info-circle\"></i></h3>\n", "<ul>\n", "\t<li>Leave both parameters as default:\n", "\t<ul>\n", "\t\t<li><strong>differential expression results</strong>: gene_scores</li>\n", "\t\t<li><strong>max number of genes to show</strong>: 20</li>\n", "\t</ul>\n", "\t</li>\n", "    <li>Click <strong><em>Run</em></strong></li>\n", "\t<li>When the table has been created, click on any links on the <q>GeneCard</q> column to learn about that particular gene</li>\n", "</ul>\n", "</div>\n"], "cell_type": "markdown"}, {"metadata": {"nbtools": {"show_code": false, "description": "Show differentially expressed genes with links to genecards.org", "type": "uibuilder", "param_values": {"actual_ranking": "false", "output_var": "", "df": "gene_scores", "max_number_of_genes_to_show": "20"}, "name": "Table Creator"}, "genepattern": {"type": "uibuilder", "output_variable": "", "param_values": {"actual_ranking": "false", "df": "gene_scores", "max_number_of_genes_to_show": "20"}, "show_code": false}, "scrolled": false}, "source": ["import numpy as np\n", "import pandas as pd\n", "\n", "def make_clickable(url):\n", "    to_display = url.split('=')[1]\n", "    return '<a href=\"{}\" target=\"_blank\">{}</a>'.format(url, to_display)\n", "\n", "def make_table(df,max_number_of_genes_to_show, actual_ranking=False):\n", "    df['Rank'] = range(1,len(df)+1)\n", "#     pd.options.display.max_colwidth = 100\n", "    if max_number_of_genes_to_show > len(df):\n", "        max_number_of_genes_to_show = len(df)\n", "        print(\"You want to show more genes than your data contains \u0ca0_\u0ca0\")\n", "        print(\"Showing you only {} instead (i.e., all the genes you have provided)\".format(max_number_of_genes_to_show))\n", "    \n", "    if actual_ranking:\n", "        df['abs-Score'] = np.absolute(df['Score'])\n", "        sorted_df = df.sort_values(by=[\"abs-Score\"], ascending=False)\n", "        sorted_df = sorted_df.head(max_number_of_genes_to_show)\n", "    else:\n", "        to_show = int(np.floor(max_number_of_genes_to_show/2))\n", "        sorted_df = df.head(to_show).append(df.tail(to_show))\n", "\n", "    sorted_df['GeneCards link']=pd.Series([\"http://www.genecards.org/cgi-bin/carddisp.pl?gene={}\".format(gene) \n", "                                     for gene in sorted_df.index.values], index=sorted_df.index)\n", "    \n", "    cols = list(sorted_df)\n", "    # move the column to head of list using index, pop and insert\n", "    cols.insert(0, cols.pop(cols.index('GeneCards link')))\n", "    # use ix to reorder\n", "    sorted_df = sorted_df.ix[:, cols]\n", "    sorted_df.set_index('Rank', inplace=True)\n", "    styled_table = sorted_df.style.format({'GeneCards link': make_clickable})\n", "    styled_table\n", "    return styled_table\n", "\n", "genepattern.GPUIBuilder(make_table, name=\"Table Creator\", \n", "                        description=\"Show differentially expressed genes with links to genecards.org\",\n", "                        parameters={\n", "                                \"df\":{\"name\": \"differential_expression_results\",\n", "                                      \"description\":'The output from differerential expression (for this exercise, leave it as \"gene_scores\")',\n", "                                      \"default\":\"gene_scores\"},\n", "                                \"max_number_of_genes_to_show\":{\"description\":\"Maximum number of genes to show in the heatmap (half will be overexpressed in one class and half in the other)\",\n", "                                                               \"default\":20},\n", "                                \"actual_ranking\":{\"hide\":True},\n", "                                \"output_var\":{\"hide\":True},\n", "                        })"], "cell_type": "code", "outputs": [{"data": {"text/plain": ["UIBuilder(description='Show differentially expressed genes with links to genecards.org', function_import='make\u2026"], "application/vnd.jupyter.widget-view+json": {"version_major": 2, "model_id": "0314af9f41cf4124a8e0f084b0ce77a1", "version_minor": 0}}, "metadata": {}, "output_type": "display_data"}], "execution_count": 3}, {"metadata": {"collapsed": true}, "source": ["<div class=\"well\">\n", "    <strong>Note:</strong> We are raking based on the score for each gene. This value can be positive (overexpressed in one class and under expressed in the other) or negative (the opposite expression profile). Hence, we want to look at high positive values but also to highly negative values.\n", "</div>"], "cell_type": "markdown"}, {"metadata": {}, "source": ["# 2. Triple Negative BRCA"], "cell_type": "markdown"}, {"metadata": {}, "source": ["<p>We will now compute differentially expressed genes of triple negative BRCA vs non triple negative BRCA. This step uses an information-theoretic method to find significantly differentially expressed transcripts or genes.</p>\n", "\n", "<p>Below you can find the files we will use.</p>\n", "\n", "<ul>\n", "    <li><a class=\"nbtools-markdown-file\" href=\"https://datasets.genepattern.org/data/TCGA_BRCA/GSEA/brca_primary_all.gct\">brca_primary_all.gct</a></li>\n", "    <li><a class=\"nbtools-markdown-file\" href=\"https://datasets.genepattern.org/data/TCGA_BRCA/GSEA/triple_negative.cls\">triple_negative.cls</a></li>\n", "</ul>"], "cell_type": "markdown"}, {"metadata": {}, "source": ["<div class=\"alert alert-info\">\n", "<h3 style=\"margin-top: 0;\"> Instructions <i class=\"fa fa-info-circle\"></i></h3>\n", "<ol>\n", "    <li>Click on the gear icon and toggle code view to hide the code of the cell below. Notice the imports, these will be automatically executed when the notebook loads</li>\n", "\t<li>For the <strong>gene expression dataset</strong> parameter, select from the dropdown menu the file:&nbsp;<strong>BRCA_dataset.preprocessed.gct</strong></li>\n", "\t<li>For the <strong>phenotype file</strong> parameter,&nbsp;select from the dropdown menu the file:&nbsp;<strong>BRCA_Dataset.cls</strong></li>\n", "\t<li>Leave the rest of the parameters as default.\n", "    <!-->\n", "\t<ul>\n", "\t\t<li><strong>output filename</strong>: diffex_output</li>\n", "\t\t<li><strong>ranking method</strong>: Pearson Correlation Matching</li>\n", "\t\t<li><strong>max number of genes to show</strong>: 20</li>\n", "\t\t<li><strong>number of permutations</strong>: 10</li>\n", "\t\t<li><strong>title</strong>: Differential Expression Results</li>\n", "\t\t<li><strong>random seed</strong>: 20180920</li>\n", "\t\t<li><strong>output variable</strong>: gene_scores</li>\n", "\t</ul>\n", "    </!-->\n", "\t</li>\n", "\t<li>Click <strong>Run</strong></li>\n", "</ol>\n", "</div>\n"], "cell_type": "markdown"}, {"metadata": {}, "source": ["<div class=\"well well-sm\">\n", "    <strong>Note:</strong> Because of the size of this dataset, running this analysis will take some time (especially the plotting portion). Remember to look at the status of your kernel represented in the dot in the upper right corner (next to \"Python 3.7\").\n", "</div>"], "cell_type": "markdown"}, {"metadata": {"nbtools": {"show_code": false, "description": "Sort genes according to their association with a discrete phenotype or class vector.", "type": "uibuilder", "param_values": {"number_of_permutations": 10, "output_var": "gene_scores", "phenotype_file": null, "output_filename": "diffex_output", "random_seed": 20121020, "ranking_method": "custom_pearson_corr", "gene_expression": null, "max_number_of_genes_to_show": 20, "title": "Differential Expression Results", "show_plot": "true"}, "name": "Differential gene expression, discrete pheotype."}}, "source": ["import genepattern\n", "import ccalnoir as ccal\n", "\n", "from ccalnoir import differential_gene_expression\n", "# import pandas as pd\n", "# import urllib.request\n", "from ccalnoir import compute_information_coefficient\n", "from ccalnoir import custom_pearson_corr\n", "RANDOM_SEED = 20180314\n", "\n", "genepattern.GPUIBuilder(differential_gene_expression, name=\"Differential gene expression, discrete pheotype.\", \n", "                        description=\"Sort genes according to their association with a discrete phenotype or class vector.\",\n", "                        parameters={\n", "                                \"gene_expression\":{\"name\":\"gene_expression_dataset\",\n", "                                                   \"type\": \"file\",\n", "                                                   \"kinds\": [\"gct\"]},\n", "                                \"phenotype_file\":{\"type\": \"file\",\n", "                                                  \"kinds\": [\"cls\"]},\n", "                                \"ranking_method\":{\n", "                                                  \"default\": \"custom_pearson_corr\",\n", "                                                  \"choices\":{'Pearson Correlation Matching':\"custom_pearson_corr\",\n", "                                                             'Information Coefficient Matching':\"compute_information_coefficient\",\n", "                                                            }\n", "                                                     },\n", "                                \"title\":{\"default\":\"Differential Expression Results\"},\n", "                                \"output_filename\":{\"default\": \"diffex_output\",},\n", "                                \"ramdon_seed\":{\"default\":20180314,},\n", "                                \"output_var\":{\"default\": \"gene_scores\",},\n", "                        })"], "cell_type": "code", "outputs": [{"data": {"text/plain": ["UIBuilder(description='Sort genes according to their association with a discrete phenotype or class vector.', \u2026"], "application/vnd.jupyter.widget-view+json": {"version_major": 2, "model_id": "407cbb8118594a76b0fb04847a944b93", "version_minor": 0}}, "metadata": {}, "output_type": "display_data"}], "execution_count": 4}, {"metadata": {}, "source": ["## Validate experimental results"], "cell_type": "markdown"}, {"metadata": {}, "source": ["We will corroborate the results summarized in the abstract of \"A New Gene Expression Signature for Triple-Negative Breast Cancer Using Frozen Fresh Tissue before Neoadjuvant Chemotherapy\" by Santuario-Facio et al https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5469719/ :  \n", "\n", "> Forty genes showed differential expression pattern in TNBC tumors. Among these, nine overexpressed genes (PRKX/PRKY, UGT8, HMGA1, LPIN1, HAPLN3, FAM171A1, BCL141A, FOXC1, and ANKRD11), and one underexpressed gene (ANX9) are involved in general metabolism. Based on this biochemical peculiarity and the overexpression of BCL11A and FOXC1 (involved in tumor growth and metastasis, respectively)[...]"], "cell_type": "markdown"}, {"metadata": {}, "source": ["# From https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5469719/\n", "genes_of_interest = ['PRKX','PRKY', 'UGT8','HMGA1','LPIN1','HAPLN3','FAM171A1','BCL141A','FOXC1','ANKRD11','BCL11A','ANX9']"], "cell_type": "code", "outputs": [], "execution_count": null}, {"metadata": {}, "source": ["def validate_results(genes_of_interest, gene_scores):\n", "    tn_genes = pd.DataFrame()\n", "    ranks = gene_scores.rank(ascending=False)['Score']\n", "    for gene in genes_of_interest:\n", "        if gene in gene_scores.index:\n", "            tn_genes = tn_genes.append(gene_scores.loc[gene])\n", "            tn_genes.loc[gene,'rank'] = ranks[gene]\n", "        else:\n", "            print(f\"Gene {gene} not present.\")\n", "    tn_genes = tn_genes.sort_values(by='Score',ascending=False)\n", "    return tn_genes\n", "\n", "validate_results(genes_of_interest, gene_scores)"], "cell_type": "code", "outputs": [], "execution_count": null}, {"metadata": {"nbtools": {"show_code": false, "description": "Show differentially expressed genes with links to genecards.org", "type": "uibuilder", "param_values": {"actual_ranking": "false", "output_var": "", "df": "gene_scores", "max_number_of_genes_to_show": 20}, "name": "Table Creator"}}, "source": ["import numpy as np\n", "import pandas as pd\n", "\n", "def make_clickable(url):\n", "    to_display = url.split('=')[1]\n", "    return '<a href=\"{}\" target=\"_blank\">{}</a>'.format(url, to_display)\n", "\n", "def make_table(df,max_number_of_genes_to_show, actual_ranking=False):\n", "    df['Rank'] = range(1,len(df)+1)\n", "#     pd.options.display.max_colwidth = 100\n", "    if max_number_of_genes_to_show > len(df):\n", "        max_number_of_genes_to_show = len(df)\n", "        print(\"You want to show more genes than your data contains \u0ca0_\u0ca0\")\n", "        print(\"Showing you only {} instead (i.e., all the genes you have provided)\".format(max_number_of_genes_to_show))\n", "    \n", "    if actual_ranking:\n", "        df['abs-Score'] = np.absolute(df['Score'])\n", "        sorted_df = df.sort_values(by=[\"abs-Score\"], ascending=False)\n", "        sorted_df = sorted_df.head(max_number_of_genes_to_show)\n", "    else:\n", "        to_show = int(np.floor(max_number_of_genes_to_show/2))\n", "        sorted_df = df.head(to_show).append(df.tail(to_show))\n", "\n", "    sorted_df['GeneCards link']=pd.Series([\"http://www.genecards.org/cgi-bin/carddisp.pl?gene={}\".format(gene) \n", "                                     for gene in sorted_df.index.values], index=sorted_df.index)\n", "    \n", "    cols = list(sorted_df)\n", "    # move the column to head of list using index, pop and insert\n", "    cols.insert(0, cols.pop(cols.index('GeneCards link')))\n", "    # use ix to reorder\n", "    sorted_df = sorted_df.ix[:, cols]\n", "    sorted_df.set_index('Rank', inplace=True)\n", "    styled_table = sorted_df.style.format({'GeneCards link': make_clickable})\n", "    styled_table\n", "    return styled_table\n", "\n", "genepattern.GPUIBuilder(make_table, name=\"Table Creator\", \n", "                        description=\"Show differentially expressed genes with links to genecards.org\",\n", "                        parameters={\n", "                                \"df\":{\"name\": \"differential_expression_results\",\n", "                                      \"description\":'The output from differerential expression (for this exercise, leave it as \"gene_scores\")',\n", "                                      \"default\":\"gene_scores\"},\n", "                                \"max_number_of_genes_to_show\":{\"description\":\"Maximum number of genes to show in the heatmap (half will be overexpressed in one class and half in the other)\",\n", "                                                               \"default\":20},\n", "                                \"actual_ranking\":{\"hide\":True},\n", "                                \"output_var\":{\"hide\":True},\n", "                        })"], "cell_type": "code", "outputs": [{"data": {"text/plain": ["UIBuilder(description='Show differentially expressed genes with links to genecards.org', function_import='make\u2026"], "application/vnd.jupyter.widget-view+json": {"version_major": 2, "model_id": "7c756427060146e59bb6c3d63e42890c", "version_minor": 0}}, "metadata": {}, "output_type": "display_data"}], "execution_count": 5}, {"metadata": {}, "source": ["# 4. UI Builder example"], "cell_type": "markdown"}, {"metadata": {}, "source": ["We can turn code we use repetitively into a function and wrap it around the GenePattern UI Builder."], "cell_type": "markdown"}, {"metadata": {}, "source": ["## Summary"], "cell_type": "markdown"}, {"metadata": {}, "source": ["<div class=\"alert alert-info\">\n", "<h3 style=\"margin-top: 0;\"> Instructions <i class=\"fa fa-info-circle\"></i></h3>\n", "<p>Run the cell below to create a UI Builder cell. Note that this cell executes the same code that we used earlier.</p>\n", "</div>"], "cell_type": "markdown"}, {"metadata": {"nbtools": {"show_code": true, "description": "", "type": "uibuilder", "param_values": {"genes_of_interest": "genes_of_interest", "gene_scores": "gene_scores"}, "name": "validate_results"}}, "source": ["@genepattern.build_ui\n", "def validate_results(genes_of_interest, gene_scores):\n", "    tn_genes = pd.DataFrame()\n", "    ranks = gene_scores.rank(ascending=False)['Score']\n", "    for gene in genes_of_interest:\n", "        if gene in gene_scores.index:\n", "            tn_genes = tn_genes.append(gene_scores.loc[gene])\n", "            tn_genes.loc[gene,'rank'] = ranks[gene]\n", "        else:\n", "            print(f\"Gene {gene} not present.\")\n", "    tn_genes = tn_genes.sort_values(by='Score',ascending=False)\n", "    return tn_genes"], "cell_type": "code", "outputs": [{"data": {"text/plain": ["UIBuilder(function_import='validate_results', name='validate_results', params=[{'name': 'genes_of_interest', '\u2026"], "application/vnd.jupyter.widget-view+json": {"version_major": 2, "model_id": "fe1e025b8e6c495da64730a578484e31", "version_minor": 0}}, "metadata": {}, "output_type": "display_data"}], "execution_count": 6}], "nbformat_minor": 1, "nbformat": 4}